### ISSUE TITLE
restic stats is unreasonably slow in comparison to other operations

### ISSUE THREAD

**Issue by cdhowie** [restic/restic#2126]
Output of `restic version`
--------------------------
```
restic 0.9.3 compiled with go1.11.1 on linux/amd64
```

How did you run restic exactly?
-------------------------------
```
restic stats --mode restore-size
```

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
Local disk.

Expected behavior
-----------------
This mode should take no longer than the part of `restic prune` that finds unused objects (about 15-20 minutes on this repository).

Actual behavior
---------------
This analysis takes multiple hours.

Do you have any idea what may have caused this?
-----------------------------------------------
My suspicion was that, while `restic prune`'s unused-object discovery mechanism prunes trees that have already been marked as used, `restic stats` on the other hand did not prune any subtrees and therefore has to recompute the size of each tree each time it is encountered.

However, looking through the source code, it seems that this is attempted?  Perhaps this code is not functioning correctly?

Do you have an idea how to solve the issue?
-------------------------------------------
Not without digging into the source code more.

**Comment by cdhowie** [comment]
Timing from `restic stats --mode restore-size`:

```
real    131m25.498s
user    123m49.552s
sys     3m59.716s
```

Timing from `restic prune` on the same repository:

```
real    14m27.800s
user    8m14.352s
sys     0m25.212s
```

**Comment by ngosang** [comment]
I can reproduce this issue in Restic 0.12.1 (Linux). For me, the issue with the `stats command` is not the time but the CPU/RAM usage.

I made a Prometheus exporter for Restic and I have to call the `stats` command multiple times in order to get the size and file count for the latest snapshots. I see there is no disk/network activity but the memory and CPU is measurable. Commands are running in the same machine the local repository is. My hardware is a RaspberryPi 4 with USB 3.0 storage. Other Restic commands run fine, at least I don't consider it's wasting resources.

![image](https://user-images.githubusercontent.com/10577978/156882618-a193bced-4ebd-4728-8394-f2695fb8c2a7.png)

**Comment by MichaelEischer** [comment]
> I see there is no disk/network activity but the memory and CPU is measurable.

No disk activity cannot be correct as the stat command has to read all the metadata for a snapshot.

I think the issue you are actually looking for is #693.

**Comment by ngosang** [comment]
Sorry, there is disk activity but it's normal. I edited my comment.
You should focus in the CPU/Memory usage, computing stats shout be cheap on those terms. For now I exporting stats to Prometheus every 30 minutes to mitigate the problem.

About the other issue, I could use that but I need the file count too.

**Comment by MichaelEischer** [comment]
> You should focus in the CPU/Memory usage, computing stats shout be cheap on those terms.

Computing the statistics will always require restic to load the repository index. And the index is responsible for the majority of the memory usage of restic. So unless we redesign that from scratch there won't be any substantial decrease in memory usage. And the CPU usage is largely determined by the directory metadata format, so this won't change either. In summary, changes to the `stat` command cannot solve the performance problems you're seeing.

> About the other issue, I could use that but I need the file count too.

There's nothing that prevents us from also saving the file count for a snapshot in addition to what was already suggested in #693.

**Comment by lenzls** [comment]
Since I was very surprised about the runtime of `stats` as well, I wanted to share my numbers:

```
# /opt/restic version
restic 0.13.1 compiled with go1.18 on linux/amd64

# time /opt/restic stats
repository f0e922b0 opened successfully, password is correct
scanning...
Stats in restore-size mode:
Snapshots processed:   407
   Total File Count:   368735348
         Total Size:   856.255 GiB

real	81m31,819s
user	87m45,828s
sys	2m5,680s
```

During the whole time I had one CPU busy with the operation. My laptop got quite hot.

**Comment by ngosang** [comment]
> There's nothing that prevents us from also saving the file count for a snapshot in addition to what was already suggested in https://github.com/restic/restic/issues/693.

It will be great if you can keep the pre-computed summary somewhere so we don't have to load the index several times.
This is my dashboards. You can get an idea about the most useful metrics from the point of view of monitoring/observability.
I have alerts if the "repository check" fails, if some client is not sync for 1 week and if some client is using more than 50GB of space.

![imagen](https://user-images.githubusercontent.com/10577978/172072727-0a68b4a8-3641-4ed0-8f15-1dee330c5f17.png)

**Comment by MichaelEischer** [comment]
The basic problem of the `stats` command is that at the moment it has to walk the whole tree for each snapshot. `prune` on the other hand only has to process each tree part once; parts that show up again in a later snapshot can be skipped.

For some of the `stats` subcommands it's probably possible to tap into the optimized `StreamTrees` function which is also used by `prune`. But I'm not sure whether that will be possible for each mode. If not, maybe there is a way to cache the intermediate statistics for tree parts which can be reused later on.

**Comment by vachanmn123** [comment]
Is there any progress on this issue? Anyone know of a way it could be made faster in any way?

**Comment by MichaelEischer** [comment]
I don't think that anyone has worked on this issue. Since restic 0.17 snapshots contain some size statistics, which can for example be read using `restic snapshots --json`.
